### PR TITLE
audit and exception on batch process

### PR DESCRIPTION
**Pull request by sankar-padakula**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/AM-1443

### Change description ###
1. CCD audit the exceptions on batch 
2. TDD iunits

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/processors/EntityWrapperProcessor.java [hunk 1]
```diff
@@ -20,31 +26,50 @@
...
+                                               .process("CCD")
+                                               .replaceExisting(false)
+                                               .roleAssignmentId(UUID.randomUUID())
+                                               .reference(ccdCaseUsers.getCaseDataId().concat(ccdCaseUsers.getUserId()))
+                                               .log(null)
+                                               .created(LocalDateTime.now())
+                                               .build();
+
+        HistoryEntity historyEntity = HistoryEntity.builder()
+                .id(requestUuid)
```
Comment (aashishaggarwal123) [inline src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/processors/EntityWrapperProcessor.java:53]: why are we keeping the history entity Id same as requestId?

File: src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/processors/EntityWrapperProcessor.java [hunk 2]
```diff
@@ -20,31 +26,50 @@
...
+                .requestEntity(RequestEntity.builder().id(requestUuid).build())
```
Comment (aashishaggarwal123) [inline src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/processors/EntityWrapperProcessor.java:54]: shouldn't this be referring to the requestEntity object created above.

File: src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/processors/EntityWrapperProcessor.java [hunk 3]
```diff
@@ -20,31 +26,50 @@
...
+                .actorId(ccdCaseUsers.getUserId())
+                .actorIdType("judge")
```
Comment (aashishaggarwal123) [inline src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/processors/EntityWrapperProcessor.java:56]: actorIdType would be always IDAM

File: src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/processors/EntityWrapperProcessor.java [hunk 4]
```diff
@@ -20,31 +26,50 @@
...
+                .roleType("judge")
```
Comment (aashishaggarwal123) [inline src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/processors/EntityWrapperProcessor.java:57]: It must be set to CASE

File: src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/processors/EntityWrapperProcessor.java [hunk 5]
```diff
@@ -20,31 +26,50 @@
...
+                .roleName("name")
+                .status("APPROVED")
+                .process("CCD")
+                .classification("org")
+                .grantType("admin")
+                .attributes(BooleanNode.getFalse())
+                .reference(ccdCaseUsers.getCaseDataId().concat(ccdCaseUsers.getUserId()))
+                .log(null)
+                .created(LocalDateTime.now())
+                .build();
```
Comment (aashishaggarwal123) [inline src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/processors/EntityWrapperProcessor.java:67]: most of these values are dummy..can we please map the correct values as per mapping.

File: src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/config/AuditSkipListener.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    @SneakyThrows
+    @Override
+    public void onSkipInWrite(EntityWrapper item, Throwable t) {
+        LOG.warn("onSkipInWrite Item " + item + " was skipped due to: " + t.getMessage());
+        AuditFaults auditFaults = new AuditFaults();
+        auditFaults.setReason(t.getMessage());
+        auditFaults.setFailedAt(AuditOperationType.WRITE.getLabel());
+        auditFaults.setCcdUsers(getJson(item.getCcdCaseUsers()));
+        auditFaults.setRequest(getJson(item.getRequestEntity()));
+        auditFaults.setHistory(getJson(item.getHistoryEntity()));
```
[the remark this answers is not in the document]
Comment (sankar-padakula) [reply src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/config/AuditSkipListener.java:59]: Yes, we have to add Live table as well. I added this in another branch code.

File: src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/writer/EntityWrapperWriter.java [hunk 7]
```diff
@@ -7,21 +7,21 @@
...
     private JdbcBatchItemWriter<RequestEntity> requestEntityWriter;
 
     @Override
     public void write(List<? extends EntityWrapper> items) throws Exception {
-        for (EntityWrapper item: items) {
-            newtableWriter.write(Collections.singletonList(item.getNewtable()));
+        for (EntityWrapper item : items) {
             requestEntityWriter.write(Collections.singletonList(item.getRequestEntity()));
+            historyWriter.write(Collections.singletonList(item.getHistoryEntity()));
         }
```
Comment (aashishaggarwal123) [inline src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/writer/EntityWrapperWriter.java:25]: are we not writing the live table record?

File: src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/processors/EntityWrapperProcessor.java (hunk 1 above)
Comment (sankar-padakula) [reply src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/processors/EntityWrapperProcessor.java:53]: This is sample builder to verify all the data is persisting or not. Nitish is doing these builder on his branch.

File: src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/processors/EntityWrapperProcessor.java (hunk 2 above)
Comment (sankar-padakula) [reply src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/processors/EntityWrapperProcessor.java:54]: This is sample builder to verify all the data is persisting or not. Nitish is doing these builder on his branch.

File: src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/writer/EntityWrapperWriter.java (hunk 7 above)
Comment (sankar-padakula) [reply src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/writer/EntityWrapperWriter.java:25]: this live table code is already available on base branch

File: src/main/resources/db/migration/V1_1__init_tables.sql [hunk 8]
```diff
@@ -72,6 +72,20 @@ CREATE TABLE role_assignment_request(
...
+	ccd_users varchar NULL,
+	request varchar NULL,
+	history varchar NULL,
+	live varchar NULL
+);
+
+
+create sequence AUDIT_ID_SEQ;
+ALTER TABLE audit_faults ALTER COLUMN id
+SET DEFAULT nextval('AUDIT_ID_SEQ');
```
Comment (aashishaggarwal123) [inline src/main/resources/db/migration/V1_1__init_tables.sql:88]: Why do we have role_assignment create table statements in this RAS batch job..This batch job is not supposed to create any table except this audit_faults

File: src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/processors/EntityWrapperProcessor.java (hunk 1 above)
Comment (aashishaggarwal123) [reply src/main/java/uk/gov/hmcts/reform/roleassignmentbatch/processors/EntityWrapperProcessor.java:53]: ok, I need to then check the main branch builder.

**Comment by sankar-padakula** [comment]
Manually merged changes to base branch

**Pull request closed without merging**